libnetwork/portallocator/portallocator.go

```diff
@@ -116,51 +116,98 @@ func (p *PortAllocator) RequestPort(ip net.IP, proto string, port int) (int, err
 // and returns that port or error if port is already busy.
 func (p *PortAllocator) RequestPortInRange(ip net.IP, proto string, portStart, portEnd int) (int, error) {
-	p.mutex.Lock()
-	defer p.mutex.Unlock()
+	if ip == nil {
+		ip = defaultIP
+	}
+	return p.RequestPortsInRange([]net.IP{ip}, proto, portStart, portEnd)
+}
 
+// RequestPortsInRange has the same behaviour as RequestPortInRange, except:
+//   - it accepts multiple addresses and attempts to find a port in the given range that's
+//     available on each of those addresses. If there is no port that's available for all
+//     addresses, no ports are reserved and an error is returned.
+//   - nil addresses are not replaced by defaultIP
+func (p *PortAllocator) RequestPortsInRange(ips []net.IP, proto string, portStart, portEnd int) (int, error) {
 	if proto != "tcp" && proto != "udp" && proto != "sctp" {
 		return 0, ErrUnknownProtocol
 	}
 
-	if ip == nil {
-		ip = defaultIP
-	}
-	ipstr := ip.String()
-	protomap, ok := p.ipMap[ipstr]
-	if !ok {
-		protomap = protoMap{
-			"tcp":  p.newPortMap(),
-			"udp":  p.newPortMap(),
-			"sctp": p.newPortMap(),
-		}
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
 
-		p.ipMap[ipstr] = protomap
+	pMaps := make([]*portMap, len(ips))
+	for i, ip := range ips {
+		ipstr := ip.String()
+		protomap, ok := p.ipMap[ipstr]
+		if !ok {
+			protomap = protoMap{
+				"tcp":  p.newPortMap(),
+				"udp":  p.newPortMap(),
+				"sctp": p.newPortMap(),
+			}
+			p.ipMap[ipstr] = protomap
+		}
+		pMaps[i] = protomap[proto]
 	}
-	mapping := protomap[proto]
+
+	// Handle a request for a specific port.
 	if portStart > 0 && portStart == portEnd {
-		if _, ok := mapping.p[portStart]; !ok {
-			mapping.p[portStart] = struct{}{}
-			return portStart, nil
+		for i, pMap := range pMaps {
+			if _, allocated := pMap.p[portStart]; allocated {
+				return 0, newErrPortAlreadyAllocated(ips[i].String(), portStart)
+			}
+		}
+		for _, pMap := range pMaps {
+			pMap.p[portStart] = struct{}{}
 		}
-		return 0, newErrPortAlreadyAllocated(ipstr, portStart)
+		return portStart, nil
 	}
 
-	port, err := mapping.findPort(portStart, portEnd)
-	if err != nil {
-		return 0, err
+	// Handle a request for a port range.
+
+	// Create/fetch ranges for each portMap.
+	pRanges := make([]*portRange, len(pMaps))
+	for i, pMap := range pMaps {
+		var err error
+		pRanges[i], err = pMap.getPortRange(portStart, portEnd)
+		if err != nil {
+			return 0, err
+		}
 	}
-	return port, nil
+
+	// Starting after the last port allocated for the first address, search
+	// for a port that's available in all ranges.
+	port := pRanges[0].last
+	for i := 0; i <= pRanges[0].end-pRanges[0].begin; i++ {
+		port++
+		if port > pRanges[0].end {
+			port = pRanges[0].begin
+		}
+
+		available := func() bool {
+			for _, pMap := range pMaps {
+				if _, allocated := pMap.p[port]; allocated {
+					return false
+				}
+			}
+			return true
+		}
+		if available() {
```

Comment:
Ok. I've been bitten by `slices` and our not-go-modules, but I guess it's ok here.

Reply:
Done.

Reply:
You were bitten by [some compiler bugs](https://go.dev/issue/66064) which have since been fixed. The `slices` and `maps` packages are all safe for us to use now.